Sources/LanguageServerProtocol/Connection.swift

```diff
@@ -136,3 +137,44 @@ extension LocalConnection: Connection {
     return id
   }
 }
+
+extension Connection {
+  /// Send the given request to the connection and await its result.
+  ///
+  /// This method automatically sends a `CancelRequestNotification` to the
+  /// connection if the task it is executing in is being cancelled.
+  ///
+  /// - Warning: Because this message is `async`, it does not provide any ordering
+  ///   guarantees. If you need to gurantee that messages are sent in-order
+  ///   use the version with a completion handler.
+  public func send<R: RequestType>(_ request: R) async throws -> R.Response {
+    let requestIDWrapper = ThreadSafeBox<RequestID?>(initialValue: nil)
+
+    @Sendable
+    func sendCancelNotification() {
+      /// Take the request ID out of the box. This ensures that we only send the
+      /// cancel notification once in case the `Task.isCancelled` and the
+      /// `onCancel` check race.
+      if let requestID = requestIDWrapper.takeValue() {
+        self.send(CancelRequestNotification(id: requestID))
+      }
+    }
+
+    return try await withTaskCancellationHandler(operation: {
+      try Task.checkCancellation()
+      return try await withCheckedThrowingContinuation { continuation in
+        let requestID = self.send(request) { result in
+          continuation.resume(with: result)
+        }
+        requestIDWrapper.value = requestID
+
+        // Check if the task was cancelled. This ensures we send a
+        // CancelNotification even if the task gets cancelled after we register
+        // the cancellation handler but before we set the `requestID`.
+        if Task.isCancelled {
+          sendCancelNotification()
+        }
+      }
+    }, onCancel: sendCancelNotification)
```

Comment:
Even though it's only used twice, IMO it would still be nice to factor this logic out similar to the function in my previous comment since it seems like it's something that's easy to get wrong. I don't feel that strongly about it though.

Reply:
You’re right. I factored it out into a separate function.